app/controllers/authentication/user_authenticator.rb

```diff
@@ -49,11 +47,20 @@ def login_path
     session_new_path
   end
 
-  def user_logged_in?(session)
-    session[:user_id].present?
+  def update_user_info(remote_ip)
+    attrs = { last_login_from: remote_ip }
+    attrs[:last_login_at] = Time.zone.now
+    attrs.merge(updatable_user_attrs) unless @user.root?
+    user.update(attrs)
```

Comment:
I think in this case it's fine to not use the bang operator.
If it somehow happens that the user entry gets invalid, this would make it impossible to log in.